src/SpaceVecAlg/PTransform.h

```diff
@@ -293,12 +293,27 @@ inline Eigen::Vector3<T> rotationError(const Eigen::Matrix3<T> & E_a_b, const Ei
 template<typename T>
 inline Eigen::Vector3<T> rotationVelocity(const Eigen::Matrix3<T> & E_a_b)
 {
-  Eigen::Vector3<T> w;
-  T acosV = (E_a_b(0, 0) + E_a_b(1, 1) + E_a_b(2, 2) - 1.) * 0.5;
+  constexpr T eps = std::numeric_limits<T>::epsilon();
+  constexpr T sqeps = details::sqrt(eps);
+  constexpr T sqsqeps = details::sqrt(sqeps);
+
+  T trace = E_a_b(0, 0) + E_a_b(1, 1) + E_a_b(2, 2);
+  T acosV = (trace - 1.) * 0.5;
   T theta = std::acos(std::min(std::max(acosV, -1.), 1.));
 
-  w = Eigen::Vector3<T>(-E_a_b(2, 1) + E_a_b(1, 2), -E_a_b(0, 2) + E_a_b(2, 0), -E_a_b(1, 0) + E_a_b(0, 1));
-  w *= sinc_inv(theta) * 0.5;
+  Eigen::Vector3<T> w(-E_a_b(2, 1) + E_a_b(1, 2), -E_a_b(0, 2) + E_a_b(2, 0), -E_a_b(1, 0) + E_a_b(0, 1));
+
+  if(1 + trace < sqsqeps) // in case of angle close to pi
+  {
+    // adapted from https://vision.in.tum.de/_media/members/demmeln/nurlanov2021so3log.pdf, sec2.2
+    Eigen::Vector3<T> s = (2 * E_a_b.diagonal() + Eigen::Vector3<T>::Constant(1 - trace)) / (3 - trace);
+    Eigen::Vector3<T> tn2 = theta * s.cwiseSqrt();
+    return (w.array() >= 0).select(tn2, -tn2);
```

Comment:
I've looked into the test failures and this is what it looks like (after adding more output in the test)
```
Failed for e: -15.5
r1v: 0.0939151  -2.04328   2.38449
r1 velocity
 -0.998213  -0.038886  0.0453796
-0.0388861  -0.153968   -0.98731
 0.0453796   -0.98731   0.152181
r1
 -0.998213   0.038886 -0.0453796
  0.038886  -0.153968   -0.98731
-0.0453796   -0.98731   0.152181
```

I'm not sure about your sign selection, based on sec 2.2 of the linked paper I tried to write as:
```cpp
if(tn2(0) > 0)
{
  if(E_a_b(0, 1) + E_a_b(1,0) < 0)
  {
    tn2(1) = -tn2(1);
  }
  if(E_a_b(0,2) + E_a_b(2, 0) < 0)
  {
    tn2(2) = -tn2(2);
  }
}
else if(tn2(1) > 0) // only if tn2(0) == 0
{
  if(E_a_b(1,2) + E_a_b(2, 1) < 0)
  {
    tn2(2) = -tn2(2);
  }
}
return tn2;
```

It solves the sign issue that is observed in the failing test output but the precision is not as good when e (in the test) is between -4 and -6

Reply:
I've pushed the modified code following our Slack conversation and clarified the branch choice w.r.t the paper

I have also extended the test for rotations around unit axes and rotation equal to pi